### PR TITLE
Split RoomMemberHandler into base and master class

### PR DESCRIPTION
**Pull request by erikjohnston**
The intention here is to split the class into the bits that can be done
on workers and the bits that have to be done on the master.

In future there will also be a class that can be run on the worker,
which will delegate work to the master when necessary.

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -31,6 +28,10 @@
 from synapse.util.async import Linearizer
 from synapse.util.distributor import user_left_room, user_joined_room
 
+import abc
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: https://github.com/matrix-org/synapse/blob/master/docs/code_style.rst:

> As per PEP-8, imports should be grouped in the following order, with a blank line between each group:
>
> standard library imports
> related third party imports
> local application/library specific imports

this is going in the opposite direction!

Comment (erikjohnston) [reply synapse/handlers/room_member.py]: Hmm. I thought we always put e.g. `import logging` last tbh. (Doesn't your IDE rearrange imports into that order?)

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -62,9 +65,66 @@ def __init__(self, hs):
         self.clock = hs.get_clock()
         self.spam_checker = hs.get_spam_checker()
 
-        self.distributor = hs.get_distributor()
-        self.distributor.declare("user_joined_room")
-        self.distributor.declare("user_left_room")
+        self.http_client = hs.get_simple_http_client()
```
Comment (erikjohnston) [inline synapse/handlers/room_member.py]: This needs to be removed as its spurious

File: synapse/handlers/room_member.py [hunk 3]
```diff
@@ -62,9 +65,66 @@ def __init__(self, hs):
...
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _user_joined_room(self, target, room_id):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _user_left_room(self, target, room_id):
+        raise NotImplementedError()
```
Comment (erikjohnston) [inline synapse/handlers/room_member.py]: These need to be docced.

File: synapse/handlers/room_member.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/room_member.py]: it does not.

File: synapse/handlers/room_member.py [hunk 4]
```diff
@@ -843,3 +836,94 @@ def _is_host_in_room(self, current_state_ids):
...
+            # capture everything from DNS failures upwards
+            #
+            logger.warn("Failed to reject invite: %s", e)
+
+            yield self.store.locally_reject_invite(
+                target.to_string(), room_id
+            )
+            defer.returnValue({})
+
+    def get_or_register_3pid_guest(self, requester, medium, address, inviter_user_id):
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: is there a reason to pass `requester` here?

File: synapse/handlers/room_member.py [hunk 5]
```diff
@@ -62,9 +65,84 @@ def __init__(self, hs):
...
 
-        self.distributor = hs.get_distributor()
-        self.distributor.declare("user_joined_room")
-        self.distributor.declare("user_left_room")
+    @abc.abstractmethod
+    def _remote_join(self, requester, remote_room_hosts, room_id, user, content):
+        """Try and join a room that this server is not in
+
+        Args:
+            remote_room_hosts (list[str]): List of servers that can be used
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: requester is missing

Comment (richvdh) [reply synapse/handlers/room_member.py]: likewise for some of the other methods

File: synapse/handlers/room_member.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/handlers/room_member.py]: Yup, as we'll need to hit out to the master and we want to include the `Requester` in the logging there.

Comment (erikjohnston) [reply synapse/handlers/room_member.py]: (We don't do it for the `user_*_room` as that just goes straight into the distributor which immediately strips the context)

**Pull request merged**